frontends/pytorch/csrc/builder/graph_importer.h

```diff
@@ -0,0 +1,117 @@
+/// least one MLIR function, but may specialize multiple versions of it,
+/// depending on usage.
+///
+/// Since graph functions typically have enough annotations for the most
+/// generic form of every type (i.e. Tensor, List, etc), and since we often
+/// want to multi-version specific specializations, we take the approach of
+/// generating a '$generic' suffixed function at that level and then generate
+/// the actual named function with using a 'numpy.generic_call' op to invoke
+/// the generic function with metadata controlling how it is legal to
+/// specialize. This leaves the process of inlining and expanding the
+/// specializations to compiler passes.
+class GraphImporter : public std::enable_shared_from_this<GraphImporter> {
+public:
+  struct MlirMappingOptions {
+    MlirContext context;
+    std::string genericFuncName;
+    std::string funcName;
+    TypeMapper &typeMapper;
+    FuncBuilder::Inserter &inserter;
+  };
+  /// Construct an importer.
+  GraphImporter(std::shared_ptr<torch::jit::Graph> graph,
+                MlirMappingOptions mappingOptions);
+
+  /// Helper to create a graph importer from a traced/scripted python function.
+  /// If the funcName of the mapping options is not set, it is set from the
+  /// function name. It is the responsibility of the caller to ensure that the
+  /// funcObj and associated graph outlives this instance.
+  static std::shared_ptr<GraphImporter>
+  forPythonJitFunc(torch::jit::Function *function,
+                   MlirMappingOptions mappingOptions);
+
+  /// Initialize for import. This is separate from the constructor purely for
+  /// ergonomics and must be called post-construction. Initialization activities
+  /// that throw go here.
+  void initialize();
+
+  /// Imports the generic function into the module.
+  void importGenericFunc();
+
+private:
+  // A scope of Graph Value * to corresponding MlirValue. Scopes nest
+  // region-wise. Note that in PyTorch, the thing called 'Block' is analagous
+  // to a capturing MLIR region.
+  struct Scope {
+    Scope() = default;
+    Scope(Scope *prev) : prev(prev) {}
+    llvm::DenseMap<torch::jit::Value *, MlirValue> valueMap;
+    Scope *prev = nullptr;
+
+    void bindValue(torch::jit::Value *torchValue, MlirValue value);
+    MlirValue findValue(torch::jit::Value *torchValue);
+    MlirValue findRequiredValue(MlirLocation loc,
+                                torch::jit::Value *torchValue);
+  };
+
+  struct Importer {
+    Importer(GraphImporter &parent, FuncBuilder *funcBuilder, MlirBlock block,
+             MlirOperation ip, Scope *scope, torch::jit::Node *node);
+    GraphImporter &parent;
+    FuncBuilder *funcBuilder;
+    MlirBlock block;
+    MlirOperation ip;
+    Scope *scope;
+    torch::jit::Node *node;
+    MlirLocation loc;
+
+    MlirContext context() { return parent.context(); }
+    void importNode();
+    void importReturnOp();
+    void importPrimNode();
```

Comment:
importPrimNode/importValueAttr seem to be internal-to-Importer helpers. But importNode/importReturnOp are "public API". Can you make that distinction clearer?